### PR TITLE
OAuth improvements (+ add user_create hook)

### PR DESCRIPTION
**Pull request by EdouardVanbelle**
Dear team, please find OAuth improvements

 * Core: Improv.: can specify an array on *_auth_type, lib will elect the best auth type according server's capabilities
 * OAuth: Refact.: migrate login flow into hooks (more evolutive code and less dependency to core code)
 * OAuth: Fix: logger prefix (include prefix during login phase)
 * OAuth: Feat: add user_create hook (map OIDC claims when creating a new user)
 * OAuth: Lean: prepare OAUTHBEARER or XOAUTH2 in Sieve, SMTP, IMAP
     
This pull request also prepare the support of OAUTHBEARER and can already be merged, I will add OAUTHBEARER once:
 - pear Net_SMTP https://github.com/pear/Net_SMTP/pull/79 is accepted + included in a new release
 - pear Net_Sieve will have a new release (thank you @alecpl for the merge of https://github.com/pear/Net_Sieve/pull/10 )
 
 Many thanks for your time

### CODE REVIEW

**Comment by Neustradamus** [comment]
@EdouardVanbelle: Excellent!

**Comment by EdouardVanbelle** [comment]
Hello I also have another commit that bring the PKCE support (security improvement)
more details: https://datatracker.ietf.org/doc/html/rfc7636

the pending change is here: https://github.com/EdouardVanbelle/roundcubemail/compare/feat/hooks-and-oauth...EdouardVanbelle:roundcubemail:feat/oauth-pkce?expand=1

Let me know it you want I add it in this PR

This will cover this request: https://github.com/roundcube/roundcubemail/issues/8757

**Comment by EdouardVanbelle** [comment]
Hello, I also implemented the use of the `nonce`, which is a security complement to `PKCE`
will be pushed on another request : https://github.com/EdouardVanbelle/roundcubemail/compare/feat/hooks-and-oauth...EdouardVanbelle:roundcubemail:feat/oauth-pkce?expand=1 )

**Comment by alecpl** [comment]
- the smaller PRs are better, I'd prefer everything auth_type related in a separate PR
- regarding auth_type, I thought it would be better to implement method auto-selection in Net_Sieve and Net_SMTP. I.e. I'd make "OAUTH" type label, when set, these packages would use "OAUTHBEARER", and if not available "XOAUTH2". Setting type as an array looks needlessly overcomplicated.

**Comment by EdouardVanbelle** [comment]
Hello I am going to split PR per features

Regarding OAUTH in libraries, I took the option that these libraries are low level
I also opt for an array, I thought it was more evolutive, seems that I took the wrong decision
I will check deeper according your suggestions ;-)

**Comment by EdouardVanbelle** [comment]
Here is the point @alecpl 

This PR is now related to code refactorization to remove dependencies in core code with Oauth class and maximise use of hooks (better evolutivity + `program/actions/login/oauth.php` part is 95% similar to `index.php > login`'s section)
Once validated I will send a  PR per features:
* oauth: user_create  (code ready)
* oauth: pkce  (security improvement) (code ready)
* oauth: nonce (security improvement) (code ready)

While waiting approval I will step back on XOAUTH2/OAUTHBEARER routing/decision that fit your suggestions

Happy end of year celebrations

File: program/include/rcmail_oauth.php [hunk 1]
```diff
@@ -67,7 +67,12 @@ class rcmail_oauth
     /** @var string */
     protected $logout_redirect_url;
 
-    /** @var array helper to map .well-known entry to config (discovery URI) */
+    // parameters used during the login phase
+    /** @var array */
```
Comment (alecpl) [inline program/include/rcmail_oauth.php:71]: `?array`, and the description should be inside /** */ here and below

File: program/include/rcmail_oauth.php [hunk 2]
```diff
@@ -1004,14 +1012,59 @@ public function managesieve_connect($options)
...
+     */
+    public function authenticate($options)
+    {
+        if (!$this->login_phase) {
+            return;
+        }
+
+        $options['user'] = $this->login_phase['username'];
+        $options['pass'] = $this->login_phase['authorization'];
+        $this->rcmail->config->set('login_password_maxlen', strlen($this->login_phase['authorization']));
```
Comment (alecpl) [inline program/include/rcmail_oauth.php:1036]: You can use shorter `$options['pass']` here.

**Review by alecpl — CHANGES_REQUESTED** [review]

File: program/include/rcmail_oauth.php (hunk 1 above)
Comment (EdouardVanbelle) [reply program/include/rcmail_oauth.php:71]: ok for the one line comment
not understood for `?array` ?

Comment (EdouardVanbelle) [reply program/include/rcmail_oauth.php:71]: ok seems to be a way to mention a hash map/array, seen in `program/lib/Roundcube/rcube_addressbook.php`

Comment (alecpl) [reply program/include/rcmail_oauth.php:71]: `?array` means `array|null`.

Comment (EdouardVanbelle) [reply program/include/rcmail_oauth.php]: Thank you for clarification, I applied correction `?array` to `$login_phase`

**Comment by EdouardVanbelle** [comment]
Hello @alecpl I added this PR: https://github.com/roundcube/roundcubemail/pull/9284 in order to fix coding style not related to this change

**Comment by Neustradamus** [comment]
@EdouardVanbelle: Good job!

Can you look for SCRAM hashes security feature too?

**Comment by EdouardVanbelle** [comment]
Hello @Neustradamus , why not, seems my dovecot supports SCRAM, will check it when I handle time to do it

**Comment by Neustradamus** [comment]
@EdouardVanbelle: Nice, thanks in advance :)

About Dovecot, please note that @stephanbosch is on SCRAM-SHA-1-PLUS/SCRAM-SHA-256-PLUS: https://github.com/dovecot/core/compare/main...stephanbosch:dovecot-core:sasl-scram-plus

He has already done several years ago several lines about SCRAM: https://github.com/search?q=repo%3Adovecot%2Fcore+scram&type=commits&s=author-date&o=desc

For -PLUS variants, there are 3 possibilities:
- tls-unique for TLS =< 1.2
- tls-server-end-point
- tls-exporter for TLS = 1.3

About Cyrus-SASL:
- https://github.com/search?q=repo%3Acyrusimap%2Fcyrus-sasl+scram&type=code

About Cyrus-Imapd:
- https://github.com/search?q=repo%3Acyrusimap%2Fcyrus-imapd+scram&type=code
- @GuidoKiener works on it: https://github.com/cyrusimap/cyrus-imapd/pull/4772

**Pull request merged**